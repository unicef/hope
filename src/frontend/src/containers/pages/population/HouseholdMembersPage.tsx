import { Box, Fade, Tooltip } from '@mui/material';
import * as React from 'react';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import {
  useHouseholdChoiceDataQuery,
  useIndividualChoiceDataQuery,
} from '@generated/graphql';
import { LoadingComponent } from '@components/core/LoadingComponent';
import { PageHeader } from '@components/core/PageHeader';
import { PermissionDenied } from '@components/core/PermissionDenied';
import { IndividualsFilter } from '@components/population/IndividualsFilter';
import { hasPermissions, PERMISSIONS } from '../../../config/permissions';
import { useBaseUrl } from '@hooks/useBaseUrl';
import { usePermissions } from '@hooks/usePermissions';
import { getFilterFromQueryParams } from '@utils/utils';
import { IndividualsListTable } from '../../tables/population/IndividualsListTable';
import { Tabs, Tab } from '@core/Tabs';
import { PeriodicDataUpdates } from '@components/periodicDataUpdates/PeriodicDataUpdates';
import { useProgramContext } from 'src/programContext';

export const HouseholdMembersPage = (): React.ReactElement => {
  const { t } = useTranslation();
  const location = useLocation();
  const { programHasPdu } = useProgramContext();
  const { businessArea } = useBaseUrl();
  const isNewTemplateJustCreated =
    location.state?.isNewTemplateJustCreated || false;

  const permissions = usePermissions();
  const { data: householdChoicesData, loading: householdChoicesLoading } =
    useHouseholdChoiceDataQuery();

  const { data: individualChoicesData, loading: individualChoicesLoading } =
    useIndividualChoiceDataQuery();

  const initialFilter = {
    search: '',
    documentType: individualChoicesData?.documentTypeChoices?.[0]?.value,
    documentNumber: '',
    admin2: '',
    sex: '',
    ageMin: '',
    ageMax: '',
    flags: [],
    orderBy: 'unicef_id',
    status: '',
    lastRegistrationDateMin: '',
    lastRegistrationDateMax: '',
  };

  const [filter, setFilter] = useState(
    getFilterFromQueryParams(location, initialFilter),
  );
  const [appliedFilter, setAppliedFilter] = useState(
    getFilterFromQueryParams(location, initialFilter),
  );

  const [currentTab, setCurrentTab] = useState(
    isNewTemplateJustCreated ? 1 : 0,
  );

  const canViewPDUListAndDetails = hasPermissions(
    PERMISSIONS.PDU_VIEW_LIST_AND_DETAILS,
    permissions,
  );

  const canViewHouseholdMembersPage = hasPermissions(
    PERMISSIONS.POPULATION_VIEW_INDIVIDUALS_LIST,
    permissions,
  );

  if (householdChoicesLoading || individualChoicesLoading)
    return <LoadingComponent />;

  if (!individualChoicesData || !householdChoicesData || permissions === null)
    return null;

  if (!canViewHouseholdMembersPage) return <PermissionDenied />;

  return (
    <>
      <PageHeader
        title={t('Household Members')}
        tabs={
          <Tabs
            value={currentTab}
            onChange={(_, newValue) => {
              setCurrentTab(newValue);
            }}
          >
            <Tab data-cy="tab-individuals" label="Individuals" />
            {programHasPdu ? (
              canViewPDUListAndDetails ? (
                <Tab
                  data-cy="tab-periodic-data-updates"
                  label="Periodic Data Updates"
                />
              ) : (
                <Tooltip title={t('Permission Denied')}>
                  <span>
                    <Tab
                      disabled
                      data-cy="tab-periodic-data-updates"
                      label="Periodic Data Updates"
                    />
                  </span>
                </Tooltip>
              )
            ) : (
              <Tooltip
                title={t(
                  'Programme does not have defined fields for periodic updates',
                )}
              >
                <span>
                  <Tab
                    disabled
                    data-cy="tab-periodic-data-updates"
                    label="Periodic Data Updates"
                  />
                </span>
              </Tooltip>
            )}
          </Tabs>
        }
      />
      <Fade in={true} timeout={500} key={currentTab}>
        <Box>
          {currentTab === 0 ? (
            <>
              <IndividualsFilter
                filter={filter}
                choicesData={individualChoicesData}
                setFilter={setFilter}
                initialFilter={initialFilter}
                appliedFilter={appliedFilter}
                setAppliedFilter={setAppliedFilter}
              />
              <Box
                display="flex"
                flexDirection="column"
                data-cy="page-details-container"
              >
                <IndividualsListTable
                  filter={appliedFilter}
                  businessArea={businessArea}
                  choicesData={householdChoicesData}
                  canViewDetails={hasPermissions(
                    PERMISSIONS.POPULATION_VIEW_INDIVIDUALS_DETAILS,
                    permissions,
                  )}
                />
              </Box>
            </>
          ) : canViewPDUListAndDetails ? (
            <PeriodicDataUpdates />
          ) : (
            <PermissionDenied />
          )}
        </Box>
      </Fade>
    </>
  );
};
