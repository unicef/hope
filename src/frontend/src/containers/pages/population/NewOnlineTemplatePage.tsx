import { BaseSection } from '@components/core/BaseSection';
import { BreadCrumbsItem } from '@components/core/BreadCrumbs';
import { LoadingComponent } from '@components/core/LoadingComponent';
import { PageHeader } from '@components/core/PageHeader';
import withErrorBoundary from '@components/core/withErrorBoundary';
import { AuthorizedUsersOnlineListCreate } from '@components/periodicDataUpdates/AuthorizedUsersOnlineListCreate';
import { FieldsToUpdateOnline } from '@components/periodicDataUpdates/FieldsToUpdateOnline';
import { FilterIndividualsOnline } from '@components/periodicDataUpdates/FilterIndividualsOnline';
import { TemplateNameOnline } from '@components/periodicDataUpdates/TemplateNameOnline';
import { useBaseUrl } from '@hooks/useBaseUrl';
import { usePermissions } from '@hooks/usePermissions';
import { useSnackbar } from '@hooks/useSnackBar';
import { Box, Button, Step, StepLabel, Stepper } from '@mui/material';
import { PDUOnlineEditCreate } from '@restgenerated/models/PDUOnlineEditCreate';
import { RestService } from '@restgenerated/services/RestService';
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { showApiErrorMessages } from '@utils/utils';
import { Formik } from 'formik';
import moment from 'moment';
import { ReactElement, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import { hasPermissions, PERMISSIONS } from 'src/config/permissions';
import { useProgramContext } from 'src/programContext';

const NewOnlineTemplatePage = (): ReactElement => {
  const { selectedProgram } = useProgramContext();
  const navigate = useNavigate();
  const location = useLocation();
  const [selected, setSelected] = useState<string[]>([]);

  const isPeople = location.pathname.includes('people');
  const { baseUrl, businessArea, programId } = useBaseUrl();
  const permissions = usePermissions();
  const { showMessage } = useSnackbar();
  const beneficiaryGroup = selectedProgram?.beneficiaryGroup;
  const { t } = useTranslation();
  const queryClient = useQueryClient();
  const { mutateAsync: createTemplateMutation } = useMutation({
    mutationFn: (params: {
      businessAreaSlug: string;
      programSlug: string;
      requestBody: PDUOnlineEditCreate;
    }) =>
      RestService.restBusinessAreasProgramsPeriodicDataUpdateOnlineEditsCreate(
        params,
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: ['periodicFields', businessArea, programId, programId],
      });
      showMessage(t('Template created successfully.'));
      navigate(
        `/${baseUrl}/population/${isPeople ? 'people' : 'individuals'}?tab=periodic-data-updates&subtab=online-edits`,
      );
    },
    onError: (error: any) => {
      showApiErrorMessages(error, showMessage, t('Failed to create template.'));
    },
  });

  const initialFilter = {
    registrationDataImportId: null,
    targetPopulationId: null,
    gender: '',
    ageFrom: '',
    ageTo: '',
    registrationDateFrom: '',
    registrationDateTo: '',
    hasGrievanceTicket: '',
    admin1: [],
    admin2: [],
    receivedAssistance: '',
  };

  const [filter, setFilter] = useState(initialFilter);
  const [checkedFields, setCheckedFields] = useState<Record<string, boolean>>(
    {},
  );

  const [activeStep, setActiveStep] = useState(0);
  const steps = [
    `Filter ${isPeople ? 'People' : beneficiaryGroup?.memberLabelPlural}`,
    'Fields to Update',
    'Authorized Users',
    'Template Name (optional)',
  ];

  const { data: periodicFieldsData, isLoading: periodicFieldsLoading } =
    useQuery({
      queryKey: ['periodicFields', businessArea, programId, programId],
      queryFn: () =>
        RestService.restBusinessAreasProgramsPeriodicFieldsList({
          businessAreaSlug: businessArea,
          programSlug: programId,
        }),
    });

  if (periodicFieldsLoading) {
    return <LoadingComponent />;
  }

  if (!periodicFieldsData) {
    return null;
  }

  const mapToInitialValuesRoundsData = (data) => {
    if (!data.results) {
      return [];
    }
    return data.results.map((item) => {
      const { name, pduData, label } = item;
      const { roundsNames } = pduData;
      return {
        field: name,
        label,
        rounds: roundsNames.map((_, roundIndex) => ({
          round: roundIndex + 1,
          roundName: roundsNames[roundIndex],
        })),
        numberOfRounds: roundsNames.length,
        roundNumber: 1,
        roundName: roundsNames[0],
      };
    });
  };

  const mappedRoundsData = mapToInitialValuesRoundsData(periodicFieldsData);

  const initialValues = {
    roundsData: mappedRoundsData,
    authorizedUsers: [],
    name: '',
  };

  const breadCrumbsItems: BreadCrumbsItem[] = [
    {
      title: beneficiaryGroup?.memberLabelPlural,
      to: `/${baseUrl}/population/${isPeople ? 'people' : 'individuals'}`,
    },
  ];

  const handleNext = () => {
    setActiveStep((prevActiveStep) => prevActiveStep + 1);
  };

  const handleBack = () => {
    setActiveStep((prevActiveStep) => prevActiveStep - 1);
  };

  // Helper functions
  const yesNoToBoolean = (value) =>
    value === 'YES' ? true : value === 'NO' ? false : null;
  const isEmpty = (value) => {
    if (value == null) return true;
    if (typeof value === 'string') return value.trim() === '';
    if (Array.isArray(value)) return value.length === 0;
    if (typeof value === 'object') {
      if (Object.keys(value).length === 0) return true;
      if ('from' in value || 'to' in value) return !value.from && !value.to;
    }
    return false;
  };

  const handleSubmit = async (values) => {
    // Build filters object
    const filters = {
      registration_data_import_id: filter.registrationDataImportId ?? null,
      target_population_id: filter.targetPopulationId ?? null,
      gender: filter.gender,
      age: {
        from: filter.ageFrom ? Number(filter.ageFrom) : null,
        to: filter.ageTo ? Number(filter.ageTo) : null,
      },
      registration_date: {
        from: filter.registrationDateFrom
          ? moment(filter.registrationDateFrom).format('YYYY-MM-DD')
          : null,
        to: filter.registrationDateTo
          ? moment(filter.registrationDateTo).format('YYYY-MM-DD')
          : null,
      },
      has_grievance_ticket: yesNoToBoolean(filter.hasGrievanceTicket),
      admin1: filter.admin1?.map((el) => el.value),
      admin2: filter.admin2?.map((el) => el.value),
      received_assistance: yesNoToBoolean(filter.receivedAssistance),
    };

    // Remove empty filters
    const filtersToSend = Object.fromEntries(
      Object.entries(filters).filter(([, value]) => !isEmpty(value)),
    );

    // Prepare rounds data
    const roundsDataToSend = values.roundsData
      .filter((el) => checkedFields[el.field])
      .map((data) => ({
        field: data.field,
        round: data.roundNumber,
        round_name: data.roundName,
        id: data.id ?? undefined, // Fix: ensure id is included if present
      }));

    const payload = {
      roundsData: roundsDataToSend,
      filters: filtersToSend,
      authorizedUsers: values.authorizedUsers,
      name: values.name || null,
    };


    try {
      await createTemplateMutation({
        businessAreaSlug: businessArea,
        programSlug: programId,
        //@ts-ignore (id not needed in payload)
        requestBody: payload,
      });
    } catch (error) {
      showApiErrorMessages(error, showMessage, t('Failed to create template.'));
    }
  };

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={handleSubmit}
      enableReinitialize
    >
      {({ values, setFieldValue, handleSubmit: formikHandleSubmit }) => {
        const roundsDataToSend = values.roundsData
          .filter((el) => checkedFields[el.field] === true)
          .map((data) => ({
            field: data.field,
            rounds: data.rounds
              .filter((round) => round.checked)
              .map((round) => ({
                round: round.round,
                round_name: round.round_name,
              })),
          }));
        return (
          <form onSubmit={formikHandleSubmit}>
            <PageHeader
              title={t('New Online Edit Page')}
              breadCrumbs={
                hasPermissions(
                  PERMISSIONS.POPULATION_VIEW_HOUSEHOLDS_LIST,
                  permissions,
                )
                  ? breadCrumbsItems
                  : null
              }
            />
            <BaseSection>
              <Stepper activeStep={activeStep} alternativeLabel>
                {steps.map((label) => (
                  <Step data-cy="label-key" key={label}>
                    <StepLabel>{label}</StepLabel>
                  </Step>
                ))}
              </Stepper>
              {activeStep === 0 && (
                <FilterIndividualsOnline
                  isOnPaper={false}
                  filter={filter}
                  setFilter={setFilter}
                />
              )}
              {activeStep === 1 && (
                <FieldsToUpdateOnline
                  values={values}
                  setFieldValue={setFieldValue}
                  checkedFields={checkedFields}
                  setCheckedFields={setCheckedFields}
                />
              )}
              {activeStep === 2 && (
                <AuthorizedUsersOnlineListCreate
                  setFieldValue={setFieldValue}
                  selected={selected}
                  setSelected={setSelected}
                />
              )}
              {activeStep === 3 && (
                <TemplateNameOnline
                  setFieldValue={setFieldValue}
                  value={values.name}
                />
              )}
              <Box
                display="flex"
                mt={4}
                justifyContent="flex-start"
                width="100%"
              >
                <Box mr={2}>
                  <Button
                    variant="outlined"
                    color="secondary"
                    component={Link}
                    to={`/${baseUrl}/population/individuals`}
                    data-cy="cancel-button"
                  >
                    Cancel
                  </Button>
                </Box>
                <Box display="flex">
                  {activeStep > 0 && activeStep < 4 && (
                    <Box mr={2}>
                      <Button
                        data-cy="back-button"
                        variant="outlined"
                        onClick={handleBack}
                      >
                        Back
                      </Button>
                    </Box>
                  )}
                  <Box>
                    {activeStep === 0 && (
                      <Button
                        variant="contained"
                        color="primary"
                        data-cy="next-button"
                        onClick={handleNext}
                      >
                        Next
                      </Button>
                    )}
                    {activeStep === 1 && (
                      <Button
                        variant="contained"
                        color="primary"
                        data-cy="next-button"
                        disabled={roundsDataToSend.length === 0}
                        onClick={handleNext}
                      >
                        Next
                      </Button>
                    )}
                    {activeStep === 2 && (
                      <Button
                        variant="contained"
                        color="primary"
                        data-cy="next-button"
                        onClick={handleNext}
                      >
                        Next
                      </Button>
                    )}
                    {activeStep === 3 && (
                      <Button
                        variant="contained"
                        color="primary"
                        data-cy="submit-button"
                        type="submit"
                      >
                        Generate Template
                      </Button>
                    )}
                  </Box>
                </Box>
              </Box>
            </BaseSection>
          </form>
        );
      }}
    </Formik>
  );
};

export default withErrorBoundary(
  NewOnlineTemplatePage,
  'NewOnlineTemplatePage',
);
