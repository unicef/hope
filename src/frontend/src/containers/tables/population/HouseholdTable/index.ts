import { HouseholdTable } from './HouseholdTable';

export { HouseholdTable };
