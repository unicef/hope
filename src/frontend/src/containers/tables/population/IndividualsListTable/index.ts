import { IndividualsListTable } from './IndividualsListTable';

export { IndividualsListTable };
