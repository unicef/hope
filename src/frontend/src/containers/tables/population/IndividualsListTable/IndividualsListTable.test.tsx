import { MockedProvider } from '@apollo/react-testing';
import * as React from 'react';
import { act } from 'react';
import wait from 'waait';
import { IndividualsListTable } from '.';
import { ApolloLoadingLink, render } from '../../../../testUtils/testUtils';
import { fakeHouseholdChoices } from '../../../../../fixtures/population/fakeHouseholdChoices';
import { fakeApolloAllIndividualsForPopulationTable } from '../../../../../fixtures/population/fakeApolloAllIndividualsForPopulationTable';

describe('containers/tables/population/IndividualsListTable', () => {
  const initialFilter = {
    search: '',
    documentType: 'individual_id',
    documentNumber: '',
    admin2: '',
    sex: '',
    ageMin: '',
    ageMax: '',
    flags: [],
    orderBy: 'unicef_id',
    status: '',
    lastRegistrationDateMin: '',
    lastRegistrationDateMax: '',
  };

  it('should render with data', async () => {
    const { container } = render(
      <MockedProvider
        addTypename={false}
        mocks={fakeApolloAllIndividualsForPopulationTable}
      >
        <IndividualsListTable
          businessArea="afghanistan"
          filter={initialFilter}
          canViewDetails
          choicesData={fakeHouseholdChoices}
        />
      </MockedProvider>,
    );
    await act(() => wait(0)); // wait for response

    expect(container).toMatchSnapshot();
  });

  it('should render loading', () => {
    const { container } = render(
      <MockedProvider
        addTypename={false}
        mocks={fakeApolloAllIndividualsForPopulationTable}
      >
        <IndividualsListTable
          businessArea="afghanistan"
          filter={initialFilter}
          canViewDetails
          choicesData={fakeHouseholdChoices}
        />
      </MockedProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
