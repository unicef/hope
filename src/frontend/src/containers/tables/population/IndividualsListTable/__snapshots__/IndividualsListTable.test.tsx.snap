// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/tables/population/IndividualsListTable should render loading 1`] = `
<div>
  <div
    class="sc-FEMpB kLrWFX"
  >
    <div
      class="sc-hRDKVd lnQjmV"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 sc-jBIHhB hbtPNr css-1gtchvp-MuiPaper-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <div
          class="MuiTableContainer-root sc-jHswkR css-70zvr5-MuiTableContainer-root"
        >
          <div
            class="sc-cSaEAk hChnVp MuiBox-root css-0"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-16wdp7y-MuiToolbar-root"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-15aaxeb-MuiTypography-root"
                data-cy="table-title"
              >
                Individuals
              </h6>
            </div>
            <div
              class="sc-cSaEAk hChnVp MuiBox-root css-1ruxp1v"
            />
          </div>
          <table
            class="MuiTable-root sc-cMqPwU sc-jZTQcj fsSCTO kxOFyj css-14k3lwa-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1o8lslm-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1e9a4b0-MuiTableCell-root"
                  data-cy="sanction-list-possible-match"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  />
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1e9a4b0-MuiTableCell-root"
                  data-cy="individual-id"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  >
                    Individual ID
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1e9a4b0-MuiTableCell-root"
                  data-cy="individual-name"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  >
                    Individual
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1e9a4b0-MuiTableCell-root"
                  data-cy="household-id"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  >
                    Household ID
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1e9a4b0-MuiTableCell-root"
                  data-cy="relationship"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  >
                    Relationship to Head of Household
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-to33az-MuiTableCell-root"
                  data-cy="individual-age"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  >
                    Age
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1e9a4b0-MuiTableCell-root"
                  data-cy="individual-sex"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  >
                    Gender
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1e9a4b0-MuiTableCell-root"
                  data-cy="individual-location"
                  scope="col"
                >
                  <span
                    class="sc-kfeOyU geVRDn"
                    data-cy="table-label"
                  >
                    Administrative Level 2
                  </span>
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
              <tr
                class="MuiTableRow-root sc-jUkaYT gWbgaU css-1o8lslm-MuiTableRow-root"
                data-cy="table-row"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium sc-deSjom edSTgE css-1dc80h3-MuiTableCell-root"
                  colspan="8"
                >
                  <span
                    class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-3sr0vz-MuiSkeleton-root"
                    style="width: 100%; height: 70px;"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="MuiTablePagination-root css-7mt0f-MuiTablePagination-root"
          data-cy="table-pagination"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-nphfye-MuiToolbar-root-MuiTablePagination-toolbar"
          >
            <div
              class="MuiTablePagination-spacer css-1wtxofq-MuiTablePagination-spacer"
            />
            <p
              class="MuiTablePagination-selectLabel css-s09cke-MuiTablePagination-selectLabel"
              id=":r6:"
            >
              Rows per page:
            </p>
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-select MuiSelect-root MuiTablePagination-input css-oegngy-MuiInputBase-root-MuiTablePagination-select"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby=":r6: :r5:"
                class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-mmygx2-MuiSelect-select-MuiInputBase-input"
                id=":r5:"
                role="combobox"
                tabindex="0"
              >
                10
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
                tabindex="-1"
                value="10"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-1gpuzre-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              class="MuiTablePagination-displayedRows css-11cfq65-MuiTablePagination-displayedRows"
            >
              0–0 of 0
            </p>
            <div
              class="MuiBox-root css-nkn1dz"
            >
              <button
                aria-label="previous page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                id=":r8:"
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="KeyboardArrowLeftIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6z"
                  />
                </svg>
              </button>
              <button
                aria-label="next page"
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                id=":r9:"
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="KeyboardArrowRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`containers/tables/population/IndividualsListTable should render with data 1`] = `
<div>
  <div
    class="sc-FEMpB kLrWFX"
  >
    <div>
      Unexpected error
    </div>
  </div>
</div>
`;
